### PR TITLE
RATIS-1994. AsyncApi send() to support optional replication level.

### PR DESCRIPTION
**Pull request by szetszwo**
See RATIS-1994

### CODE REVIEW

**Comment by jojochuang** [comment]
Thanks @szetszwo. The PR looks reasonable to me but there are a few test failures that need some attention.

**Comment by jojochuang** [comment]
The test passed locally so I suppose it's unrelated. +1

**Comment by szetszwo** [comment]
> The test passed locally so I suppose it's unrelated. ...

Yes, it is not related to this change.

It looks like that the `RaftServerImpl.close()` method was called  in the of `LeaderStateImpl.stepDown(..)`.  Synchronize the `close()` method should be able to fix it.

```java
Caused by: java.lang.NullPointerException: leaderState == null
	at org.apache.ratis.server.impl.RoleInfo.shutdownLeaderState(RoleInfo.java:90)
	at org.apache.ratis.server.impl.RaftServerImpl.changeToFollower(RaftServerImpl.java:579)
	...
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 1]
```diff
@@ -86,20 +65,7 @@
...
-import org.apache.ratis.protocol.exceptions.ReadException;
-import org.apache.ratis.protocol.exceptions.ReadIndexException;
-import org.apache.ratis.protocol.exceptions.ReconfigurationInProgressException;
-import org.apache.ratis.protocol.exceptions.ResourceUnavailableException;
-import org.apache.ratis.protocol.exceptions.ServerNotReadyException;
-import org.apache.ratis.protocol.exceptions.SetConfigurationException;
-import org.apache.ratis.protocol.exceptions.StaleReadException;
-import org.apache.ratis.protocol.exceptions.StateMachineException;
-import org.apache.ratis.protocol.exceptions.TransferLeadershipException;
+import org.apache.ratis.protocol.exceptions.*;
```
Comment (TaiJuWu) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:68]: Is there any rule to decide to use * or import individual?
In this PR, you also modified 
```
import org.apache.ratis.proto.RaftProtos.*;
```
to
```
import org.apache.ratis.proto.RaftProtos.DataStreamRequestTypeProto;
import org.apache.ratis.proto.RaftProtos.ForwardRequestTypeProto;
import org.apache.ratis.proto.RaftProtos.MessageStreamRequestTypeProto;
....
```

Comment (szetszwo) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:68]: This was to [Get around RaftServerImpl.java File length is 2,019 lines (max allowed is 2,000](https://github.com/apache/ratis/pull/1006/commits/8cc04ea97b5621407814f3aab317ca148769a54f) .  Filed RATIS-1996 to refactor RaftServerImpl.java

Honestly, I feel people focus too much on checkstyle issues.  We are better to put our attention on the code logic instead.

**Review by SzyWilliam — APPROVED** [review]
+1 the change looks good to me.

I was thinking (dangerously), If we allow a low replication level (e.g.1), we can turn ratis into some async data replication service ;)

**Comment by szetszwo** [comment]
> ... If we allow a low replication level (e.g.1), we can turn ratis into some async data replication service ;)

This is an interesting idea!  The leader may response once it has persisted the request.  However, the leader cannot yet apply it to the state machine.  This is okay for some stateless servers such as a data replication service as you mentioned..

@jojochuang , @SzyWilliam , thanks a lot for reviewing this!

**Pull request merged**